### DISCUSSION TITLE
Static files - Where should I place images and fonts?

### DISCUSSION THREAD

**Discussion by sky-xo** [uicrooks/shopify-theme-lab#43] [Q&A] [answered]
Where should I place static files like images and fonts?

And how do I make sure that images used in liquid templates get placed in the `dist/` folder? I assumed placing them in a folder like `src/images/` would be fine, but webpack isn't adding them to `dist/` because they aren't being used in a css file.

Also, placing fonts in `src/` seems to be causing an error with webpack, but I'll post the details of that in a separate issue.

By the way, great project, super handy. Thanks for making it. :-)

**Comment by sky-xo** [comment]
My current workaround is adding this to `webpack.common.js`:

```
    new CopyPlugin({
      patterns: [
        {
          from: "src/images/**/*",
          to: '[name].[ext]'
        },
      ],
    }),
```

**Comment by sergejcodes** [comment, marked as answer]
Hi Jon 👋.

You should place the images inside `shopify/assets/` directory. There is no need to process them through webpack. Add the `static` keyword to their file name e.g. `file-name.static.png`. That way they won't be cleaned by webpack during builds and git will track them.

To include them in a `.liquid` file use:
```liquid
<img src="{{ 'file-name.static.png' | asset_url }}">

<!-- or -->

<img src="{{ 'file-name.static.png' | asset_img_url: '300x' }}">
```

**Reply by sky-xo** [reply]
Awesome, thank you Sergej! 

However, this doesn't seem to work for fonts in `.css` files, only for images in `.liquid` files.

If I try to import a font into CSS using:

```
@font-face {
  font-family: "Unica77";
  src: url("Unica77-Light.static.woff") format("woff");
  font-weight: 300;
}
```

I receive the following error:

```
Error: Can't resolve 'Unica77-Light.static.woff' in '/Users/<username>/code/<project-name>/src/css'
```

It's looking for the file in `src/` instead of `shopify/assets/`. Does this mean I have to run fonts through webpack for now?

**Reply by sergejcodes** [reply]
I just did some testing, the simplest solution is:

1. In `.config/webpack/webpack.common.js` add the options object to the file-loader:

Instead of:
```js
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/,
        use: [
          'file-loader'
        ]
      }
```

It should be:
```js
      {
        test: /\.(woff|woff2|eot|ttf|otf)$/,
        use: {
          loader: 'file-loader',
          options: {
            name: '[name].[ext]',
            publicPath: './',
            outputPath: './'
          }
        }
      }
```
The font files will end up in the same directory as the final css file, so it will transform every font-face url() path

2. Also in `.config/webpack/webpack.common.js` simplify the `CleanWepackPlugin` static glob selector, there seems to be an issue with fonts:

form:
```js
    new CleanWebpackPlugin({
      cleanOnceBeforeBuildPatterns: ['**/*', '!+(*)static+(*)'],
    }),
```

to:
```js
    new CleanWebpackPlugin({
      cleanOnceBeforeBuildPatterns: ['**/*', '!*static*'],
    }),
```

3. Place your fonts somewhere in the `src/` dir e.g. `src/fonts/Unica77-Light.woff`

4. Import them like you already wrote:
```css
@font-face {
  font-family: "Unica77";
  src: url("../fonts/Unica77-Light.woff") format("woff");
  font-weight: 300;
}
```

this should fix your problem

**Reply by sky-xo** [reply]
@sergejcodes amazing, thank you so much for the detailed reply and support!

**Reply by ScottJr** [reply]
It might be worth including `.static` for assets in the docs.

**Reply by sergejcodes** [reply]
It's already mentioned in the docs. I included an additional `.static` file in the `shopify/assets/` directory for the latest release. That way the `.static` concept might be easier to spot.

**Reply by CaseJnr** [reply]
Im using shopify-theme-lab version 2.6.0 and am having issues loading fonts in development. I have followed the instructions above and placed the font files in src/fonts and set the paths to via `src: url('../fonts/fancy.woff')`.

`npm run start` adds the css to `<head>` and the bundle.css returns a 404 (as expected). If I navigate to a product page, say `https://mystore.shopify.com/products/example` the font returns a 404 as the font path is converted to `src: url(./fancy.woff)` and therefore the font is trying to be retrieved from `https://mystore.shopify.com/products/fancy.woff`.

Hows do I get fonts loading in development?

**Reply by sergejcodes** [reply]
Just tested a couple of things and found a better way of including custom fonts.
You don't need webpack as they are treated as static files (example without Tailwind):

1. add your fonts to the `shopify/assets/` directory with the `static` keyword e.g.:
`indie-flower-v12-latin-regular.static.woff2`
`indie-flower-v12-latin-regular.static.woff`

2. create a Shopify snippet `shopify/snippets/fonts.liquid` with the following content:
```liquid
{% style %}
  @font-face {
    font-family: 'Indie Flower';
    font-style: normal;
    font-weight: 400;
    src: url('{{ 'indie-flower-v12-latin-regular.static.woff2' | asset_url }}') format('woff2'),
         url('{{ 'indie-flower-v12-latin-regular.static.woff' | asset_url }}') format('woff');
  }
{% endstyle %}
```

3. render the `fonts` snippet above/below the `bundle.css` in `shopify/layout/theme.liquid`
```liquid
...
{{ 'bundle.css' | asset_url | stylesheet_tag }}
{% render 'fonts' %}
...
```

4. Create a css file and import it into `src/css/main.css` with the following content:
```css
body {
  font-family: 'Indie Flower', serif;
}
```

**Reply by CaseJnr** [reply]
Thanks for getting back to me @sergejcodes, I really appreciate it.

I ended up moving the file-loader from `webpack.common.js` to `webpack.prod.js` and updating the test value for all image and font extensions.

```
{
        test: /\.(png|svg|jpg|gif|woff|woff2|eot|ttf|otf)$/,
        use: [
          {
            loader: 'file-loader',
            options: {
              name: '[name].[ext]',
              publicPath: './',
              outputPath: './'
            }
          }
        ]
      },
```

Then I added the following url loader to `webpack.dev.js `

```
{
        test: /\.(png|svg|jpg|gif|woff|woff2|eot|ttf|otf)$/,
        use: [
          {
            loader: 'url-loader',
            options: {
              limit: 9999999999,
              name: 'pack-[name].[ext]'
            }
          }
        ]
      },
```

This allows all fonts and images to remain in assets as all paths will be converted to base64 values for development. I added a @assets alias to webpack.common.js to assist with the imports.

`'@assets': path.resolve(__dirname, '../../shopify/assets')`

Fonts and images can simply be imported with 

src: url('@assets/myfont.woff') 

I also updated `CleanWebpackPlugin` to only check for files that will be generated by webpack so the static prefix isn't required. 
 
 `cleanOnceBeforeBuildPatterns: ['bundle.js', 'bundle.css', 'pack-*.*']`
